BUG: Added check for NULL data in ufuncs

madphysicist opened the pull request:
Closes #14613. Supersedes and invalidates #20686.

Out of the two solutions (this and #20686), this one is clearly preferable. While this part of the C API is "soon" to be deprecated, it appears prominently in the tutorial, so may as well be correct.

I am not very familiar with the C-API, so the changes you see may be incomplete. My methodology was to run `grep -rnH -- '->data\>'` on the main repository, and update anything that looked like a reference to a ufunc's `data` field. I also ran `grep -rnH -- '.data\>'` and `grep -rnHF loopdata` for good measure, but that yielded no meaningful results, as expected.

This PR is not marked as WIP, but someone that actually knows what they are doing should probably look at it.

madphysicist committed:
BUG: Added check for NULL data in ufuncs

seberg commented:
I am happy with this change, but could you make sure it is documented that this can be NULL now?

madphysicist commented:
@seberg This is in response to the documentation specifically saying it can be NULL and @WarrenWeckesser and I running into the fact that it can't. I will immediately close the alternative PR #20686 if this gets merged, since it was changing the documentation to *remove* the explicit statement that the parameter could be NULL.

madphysicist commented:
@seberg Quoting from here: https://numpy.org/doc/stable/reference/c-api/ufunc.html#c.PyUFunc_FromFuncAndData:

> **data** – Should be `NULL` or a pointer to an array of size *ntypes*. This array may contain arbitrary extra-data to be passed to the corresponding loop function in the func array.

seberg commented:
Sorry, I was thinking of the UFunc struct, not the function for registration.

madphysicist committed:
DOC: Made NULL refs more explicit

madphysicist commented:
@seberg. Thanks for the catch. I updated the page here: https://numpy.org/doc/stable/reference/c-api/ufunc.html, though it's a reference to the function argument, not the field. I don't see any other references to `data` in the obvious documentation. Since you have a positive reaction to this PR, I also updated  https://numpy.org/doc/stable/user/c-info.ufunc-tutorial.html, but in #20683

seberg commented:
Thanks, this seems good to me, so going to merge later.  The only change possible change would be to add a `.. versionchanged::`, but not sure it matters here, so just going to merge later probably.
(I honestly hope it won't really ever be useful, because of new API usage kicking in before this has a chance to be used well...)

seberg commented:
Yeah, but implementation trumps documentation and if someone wants to use the NULL thing they could be surprised if their stuff doesn't work on older NumPy versions suddenly.

seberg commented:
I don't really think you need the "to conform to documented behavior" part, but sure.  I don't want to bike-shed this...  The bigger change in my view is that the slot can now be NULL (rather than an array of NULLs), but since this will only kick in for custom ufuncs somewhere out there, I am fine with that change.

(And hopefully nobody really fetches it anyway except for their own functions.)

madphysicist committed:
DOC: Added ..versionchanged:: tag

seberg commented:
LGTM, lets put this in, thanks @madphysicist!

I removed the backport candidate label, since it doesn't seem like a particular important bug fix and rather more like an enhancement.  But I don't mind if this is backported.
